### DISCUSSION TITLE
rowcount with ORM .returning() on SQLAlchemy 2.0

### DISCUSSION THREAD

**Discussion by RazerM** [sqlalchemy/sqlalchemy#12095] [Usage Questions]
On SQLAlchemy 1.4 I could use `session.execute` with a RETURNING clause on postgresql and still get a `.rowcount` back.

In [Getting Affected Row Count from UPDATE, DELETE](https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#getting-affected-row-count-from-update-delete) I see this (emphasis mine):

> The [CursorResult](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult) class is a subclass of [Result](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result) which contains additional attributes that are specific to the DBAPI cursor object. An instance of this subclass is returned when a statement is invoked via the [Connection.execute()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.execute) method. When using the ORM, the [Session.execute()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.execute) method **returns an object of this type for all INSERT, UPDATE, and DELETE statements**.

When I add `.returning(...)` I start getting a `ChunkedIteratorResult` instead which doesn't have rowcount.

**So my question is:** is this a docs bug or an actual bug? Should I be using `session.connection()` here?

It would be nice if this could work.

MCVE that works on 1.4 but not 2.0:

```python
from sqlalchemy import Column, Integer, String, create_engine, update
from sqlalchemy.orm import Session, declarative_base

Base = declarative_base()

class Book(Base):
    __tablename__ = 'books'

    id = Column(Integer, primary_key=True)
    title = Column(String, nullable=False)

engine = create_engine('postgresql:///sqlalchemytest', echo=True, future=True)

Base.metadata.create_all(engine)

with Session(engine, future=True) as session:
    session.add(Book(title="Title"))
    result = session.execute(update(Book).values(title="New title").returning(Book.id))
    print(result.scalars().all())
    print(result.rowcount)
```

**Comment by zzzeek** [comment]
I would use .connection here.    those docs should definitely note that "if RETURNING is not used", there's really no guarantee of .rowcount if RETURNING is in play since that entirely changes how the result is fetched.

**Reply by RazerM** [reply]
interestingly, the code in the example autoflushes when `session.execute(update(Book)...)` is used, but neither of these autoflush:

```python
        session.add(Book(title="Title"))
        result = session.execute(update(Book.__table__).values(title="New title").returning(Book.id))
```

```python
        session.add(Book(title="Title"))
        result = session.connection().execute(update(Book).values(title="New title").returning(Book.id))
```

my intuition is that the first one _would_ autoflush and the second one wouldn't because `session.connection()` is an opt out of the ORM stuff.

**Reply by zzzeek** [reply]
first one doesnt autoflush because it's not an ORM update statement, its a Core statement, so it's the same kind of opt out.

the second one is a mixed bag since it's connection(), so autoflush is out of the picture, but the statement still has ORM compilation aspects
